More conservative fix for issue 313 & 314

9rnsr opened the pull request:
[Issue 313](https://d.puremagic.com/issues/show_bug.cgi?id=313) - [module] Fully qualified names bypass private imports
[Issue 314](https://d.puremagic.com/issues/show_bug.cgi?id=314) - [module] Static, renamed, and selective imports are always public

Unlike #2256, this PR will keep existing name lookup rule and import behaviors that is unrelated to the issues.

Two main points:
1. Most of import declarations will insert their fully qualified names in local symbol table.

Currently, it is stored only in one place - `Module::modules`. After this change, in addition to that, each `ScopeDsymbol::symtab` will have the "local package tree" which represents the scope local imports.

For example:

``` d
class C {
    import core.stdc.stdio;
    import core.stdc.stdlib : malloc;
    import core.stdc.stdlib : free;
    void func() {
        import core.stdc.string;
    }
}
```

`symtab` of class `C` and `C.func` will have following package object trees:

```
// in C
Package('core') <----------+
 + Package('stdc') <-------|--+
    + Import('stdio')      |  |
    + Import('stdlib')     |  |
        names = ["malloc"] |  |
  +---- overnext           |  |
  |                        |  |
  +-> Import('stdlib')     |  |
        names = ["free"]   |  |
                           |  |
// in C.func()             |  |
Package('core') aliassym --+  |
 + Package('stdc') aliassym -+
    + Import('string')
```

In there, `Import` object will be the leaf of the tree.

It's done by `Import::addPackage`.
1. A part of local package tree will be merged into other tree.

This example case of public import should work.

``` d
module a;
public import std.stdio;
```

``` d
module b;
import a;
import std.algorithm;
std.stdio.writeln();    // point
```

In module b, 'std.stdio' should be visible. To realize it, import declarations will pull the fully qualified name of public imports in the imported module. Finally, module b will have the package tree:

```
Package('std')
 + Package('algorithm')
 + Package('stdio')   <--- Pulled from module a, by `import a;`
```

Same work is necessary on class inheritance. Public/protected imports in base class scope should be pulled in the derived class scope.

quickfur commented:
This is a gigantic wart in D's module system that has been plaguing us for at least 5 years; can we pretty please move forward with this?

MartinNowak commented:
> It is blocked on @WalterBright AFAICU

No, it's blocked by the lack of an implementation that doesn't change lookup rules.

MartinNowak commented:
> Isn't the fix supposed to change lookup rules pretty much by description of the issue, huh?

A fix to 314 should only change protection for selective imports but not touch the alias that's added.

9rnsr commented:
Updated for ddmd build.

MartinNowak commented:
Thanks Kenji, any chance we can split off the uncontroversial 313 change?

9rnsr commented:
@MartinNowak Sorry I can't. To fix issue 313, we need to *replace* current problematic symbol search mechanism. Therefore I can't split the fix into multiple incremental changes so each partial changes will interfere with the existing mechanism.

You can observe the test case changes, or pull this into a local repo to build and test the new compiler behavior.

If you ask about code change, I'll answer to it.

MartinNowak commented:
I was under the impression that this pull still removed the aliases of selective imports, but that is not the case. So there isn't anything controversial in this pull, despite the implementation.

MartinNowak commented on src/mtype.c:
```diff
@@ -6516,10 +6516,23 @@ void TypeQualified::resolveHelper(Loc loc, Scope *sc,
-                s = si;
+                if (idents.dim == 0)
+                {
+                    /* Special case for runnable/Same.d:
```
What's extremely weird is (a hidden feature?) that this works with base classes, but not with functions or when directly using the symbol.
```d
module foo;
void foo() {}
```
```
module bar;
import foo;
void test() { foo(); } // Error: function expected before (), not module foo of type void
pragma(msg, foo); // Error: cannot interpret module foo at compile time
```

9rnsr replied:
Yes, (maybe) it's very weird feature for the symbols parsed as types. I didn't know about that when I noticed.

MartinNowak replied:
Maybe some Javaish idea (one class per module) or simply a bug.

9rnsr replied:
Ah, indeed it works like Java.

MartinNowak commented:
Had an idea how to avoid all those trees.
If you have to resolve a qualified symbol `a.b.c.d.e`, one could try to find the longest prefix in a plain hashtable, e.g. lookup import `a.b.c.d`, if that fails, try `a.b.c`. Then lookup the remaining suffix in that import.
Not sure if it would really simplify the implementation.

9rnsr commented:
> Not sure if it would really simplify the implementation though.

It might reduce the search cost, but definitely it would increase the code complexity. To do it, compiler need to have two distinct search mechanism - hashing for the longest fully qualified name, and hashing for each scope levels.

MartinNowak commented:
- [ ] https://github.com/higgsjs/Higgs/pull/186
- [ ] https://github.com/rejectedsoftware/vibe.d/pull/1066 
- [ ] https://github.com/D-Programming-Language/dub/pull/544
- [x] [dub-registry](https://github.com/D-Programming-Language/dub-registry)

One thing I found, that might be controversial.
```d
import std.ascii : isWhite;
// static import std.ascii : isWhite; // Error: static import std cannot have an import bind list

bool foo()
{
    return std.ascii.isWhite(' '); // no longer works
}
```
You mentioned in the other pull that combining renamed/static/selective requires some kind of code breakage. What are the plans for this area?

9rnsr commented:
> You mentioned in the other pull that combining renamed/static/selective requires some kind of code breakage. What are the plans for this area?

It's purely an enhancement. After this merge, I can implement it easily.

MartinNowak commented:
Is the code above supposed to work?

9rnsr commented:
> Is the code above supposed to work?

No, because this PR is a bug-fix.

MartinNowak commented:
> No, because this PR is a bug-fix.

I see, it doesn't work in 2.067.0 either. Let's figure why it used to work in vibe.d though.
https://github.com/rejectedsoftware/vibe.d/pull/1066/files#diff-3783ec0fb4e4ec8b36cd4c08c01da911L680

9rnsr commented:
> https://github.com/rejectedsoftware/vibe.d/pull/1066/files#diff-3783ec0fb4e4ec8b36cd4c08c01da911L680

That's just the appearance of issue 313. If a module `x` imports `std.ascii` (even if it's imported privately) in module scope, it would make the fully qualified name globally visible. Then if you import the module `x`, the fully qualified package name could be visible in your module. In 2.067, many of module level imports in Phobos was changed to scope local imports. The Phobos changes have being invisible some fully qualified Phobos package names in your modules, as you see.

MartinNowak commented:
> That's just the appearance of issue 313. If a module x imports std.ascii

I see makes sense. I'll try to review this pull in more detail during the next week.
Also mailed Walter and Andrei to give their OK for fixing this in 2.068.0.

MartinNowak commented:
> Currently, it is stored only in one place - Module::modules. After this change, in addition to that, each ScopeDsymbol::symtab will have the "local package tree" which represents the scope local imports.

Can we properly deprecate usage of FQN names instead of making them a hard fault?
If the global module tree still exists, we could use that as a fallback lookup with the deprecation warning attached.

9rnsr commented:
> Can we properly deprecate usage of FQN names instead of making them a hard fault?

Issue 313 and 314 are bugs. And I hate to abuse deprecation message for any bug migrations. Even if it's possible, I won't work for that.

MartinNowak commented:
I don't like that either, but we need to deal with the fallout of our changes.
Can we at least print a detailed error message, saying you need to `import core.stdc.stdio;`?

MartinNowak commented:
Got an OK, let's tackle that for 2.068.0. We need to sort out any implementation concerns and mitigate update pains. Personally I'd be fine with a good error message, but given how much code this affects we should go the warning/deprecation path.

MartinNowak commented:
From Andrei.

> We could then go with -issue313 or something. My thoughts on the matter:

> 1. It's a hole straight in the middle of D's modularity, the kind that will hit harder with larger projects. The PR leakage resulting from this may also be deadly. We must fix it.

> 2. It'll produce lots of annoyance in random places for random projects that currently just work as expected, even if for wrong reasons.

> Reminds me of the bug Walter fixed a while ago regarding template instantiation. Suddenly people got linker errors for subtle reasons.

> That's why I'm thinking an opt-in angle would be best.

I think the usual warning+deprecation makes more sense than a switch that splits the user base, think -property.
@Dicebot could you work on this before we merge it?

You can find perfect test cases in the bug reports and in this pull.
https://issues.dlang.org/show_bug.cgi?id=313#c0
https://issues.dlang.org/show_bug.cgi?id=314#c0

MartinNowak commented:
Thanks, let's do this. Ideally it would get ready a few weeks before the release. Shout out if you need some help.

MartinNowak commented:
> What is the planned schedule for 2.068 btw?

Next month, branching in ~2 weeks.

MartinNowak commented:
Sigh, endlessly delayed.

9rnsr commented:
Rebased on the latest master.

MartinNowak commented:
> 1. Most of import declarations will insert their fully qualified names in local symbol table.
> Currently, it is stored only in one place - Module::modules. After this change, in addition to that, each ScopeDsymbol::symtab will have the "local package tree" which represents the scope local imports.

How about keeping one global package tree and using bitvectors in each scope to mask the modules?

9rnsr commented:
> How about keeping one global package tree and using bitvectors in each scope to mask the modules?

It would be more efficient, but one problem I can see is it's difficult to implement. Currently the global package tree is implemented by `DsymbolTable`, which using AA. So, there's no obvious order between the sub package names, and the bit position of masked package is hard to determine.

I don't want to bloat size of this PR.

9rnsr commented:
Ready for ddmd. And now, a big commit is divided based on the change of logic.

MartinNowak commented:
> Ready for ddmd. And now, a big commit is divided based on the change of logic.

Let's try to get this into 2.070. It's a bit too big to mix with the ddmd transition IMHO.

9rnsr committed:
Add DsymbolTable parameter to Package.resolve() to handle local package tree

9rnsr committed:
Package now can become an alias to Module or Import

9rnsr committed:
Build local package tree in Import.addMember

9rnsr committed:
Distinguish public imports in module/class members

In a derived scope of module (== the scope where it's imported) or class
(the member scope of its derived class), the full package name of those
public imports should be visible. To do that, those FQNs will be pulled
into the local package tree of the derived scope.

9rnsr committed:
Fix search functions and its callers not to see invisible symbols

9rnsr committed:
Fix issue 313 & 314 done - explanation commit log and add/fix test cases

Points:

1. Most of import declarations will insert their fully qualified names in local symbol table.

Currently, it is stored only in one place - `Module.modules`. After this change, in addition to that,
each `ScopeDsymbol.symtab` will have the "local package tree" which represents the scope local imports.

For example:

```d
class C {
    import core.stdc.stdio;
    import core.stdc.stdlib : malloc;
    import core.stdc.stdlib : free;
    void func() {
        import core.stdc.string;
    }
}
```

`symtab` of class `C` and `C.func` will have following package object trees:

```
// in C
Package('core') <----------+
 + Package('stdc') <-------|--+
    + Import('stdio')      |  |
    + Import('stdlib')     |  |
        names = ["malloc"] |  |
  +---- overnext           |  |
  |                        |  |
  +-> Import('stdlib')     |  |
        names = ["free"]   |  |
                           |  |
// in C.func()             |  |
Package('core') aliassym --+  |
 + Package('stdc') aliassym -+
    + Import('string')
```

In there, `Import` object will be the leaf of the tree.
It's done by `Import.addPackage`.

2. A part of local package tree will be merged into other tree.

This example case of public import should work.

```d
module a;
public import std.stdio;
```
```d
module b;
import a;
import std.algorithm;
std.stdio.writeln();    // point
```

In module b, 'std.stdio' should be visible. To realize it, import declarations will pull
the fully qualified name of public imports in the imported module.
Finally, module b will have the package tree:

```d
Package('std')
 + Package('algorithm')
 + Package('stdio')   <--- Pulled from module a, by `import a;`
```

Same work is necessary on class inheritance. Public/protected imports in base class scope
should be pulled in the derived class scope.

9rnsr committed:
fix Issue 313 - [module] Fully qualified names bypass private imports

9rnsr committed:
fix Issue 314 - [module] Static, renamed, and selective imports are always public

9rnsr committed:
fix Issue 12735 - imports via mixin template become public imports

9rnsr committed:
Add asymmetric behavior of function local imports for backward compatibility

See also: Issue 15400 - inconsistent local `import` behavior

tsbockman commented:
> Let's try to get this into 2.070. It's a bit too big to mix with the ddmd transition IMHO.

I just stumbled across this bug today. It's very annoying; I hope this PR is still on the agenda for 2.070.

JackStouffer commented:
@MartinNowak what's the likelihood of this getting in the next release?

adamdruppe commented:
2.070 just got tagged beta...

ibuclaw commented:
@MartinNowak - is this on the cards?  Could this be a candidate for pulling into 2.071 as soon as development window opens?

MartinNowak commented:
> @MartinNowak - is this on the cards? Could this be a candidate for pulling into 2.071 as soon as development window opens?

I don't think we should pull this PR as is. It's a fairly complex implementation (see the [diagram](https://github.com/D-Programming-Language/dmd/pull/3407#issue-30435236) and a proposal for a [simpler 314 solution](https://github.com/D-Programming-Language/dmd/pull/3407#issuecomment-129380687)).
Also this PR doesn't contain any backwards compatibility shims, it'll just unnecessarily break lots of code.
I think we should just use 2.071 and hop on fixing both of those bugs properly.

ibuclaw commented:
> It's a fairly complex implementation 

I don't think the diagram does justice to this PR, although I have a faint memory that says the current implementation isn't so dissimilar in design either.

>  I'd even volunteer to implement the 313 fix.

OK.
